pkg/tbtc/signing_test.go

```diff
@@ -202,6 +342,252 @@ func TestSigningRetryLoop(t *testing.T) {
+			broadcastingMembersIndexes: []group.MemberIndex{1, 3, 5},
+			expectedErrors:             make(map[group.MemberIndex]error),
+			expectedResults: map[group.MemberIndex][]group.MemberIndex{
+				1: {1, 3, 5},
+				3: {1, 3, 5},
+				5: {1, 3, 5},
+			},
+		},
+		"minority of members broadcasted announcements": {
+			message:                    big.NewInt(300),
+			broadcastingMembersIndexes: []group.MemberIndex{1, 3},
+			expectedErrors: map[group.MemberIndex]error{
+				1: fmt.Errorf("ready members count is lesser than the honest threshold"),
+				3: fmt.Errorf("ready members count is lesser than the honest threshold"),
+			},
+			expectedResults: make(map[group.MemberIndex][]group.MemberIndex),
+		},
+	}
+
+	for testName, test := range tests {
+		t.Run(testName, func(t *testing.T) {
+			broadcastChannel, err := localProvider.BroadcastChannelFor(
+				test.message.Text(16),
+			)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			membershipValidator := group.NewMembershipValidator(
+				&testutils.MockLogger{},
+				signingGroupOperators,
+				localChain.Signing(),
+			)
+
+			announcer := newBroadcastSigningAnnouncer(
+				chainConfig,
+				broadcastChannel,
+				membershipValidator,
+			)
+
+			resultsChan := make(
+				chan *memberResult,
+				len(test.broadcastingMembersIndexes),
+			)
+			errorsChan := make(
+				chan *memberError,
+				len(test.broadcastingMembersIndexes),
+			)
+
+			wg := sync.WaitGroup{}
+			wg.Add(len(test.broadcastingMembersIndexes))
+
+			for _, broadcastingMemberIndex := range test.broadcastingMembersIndexes {
+				go func(memberIndex group.MemberIndex) {
+					defer wg.Done()
+
+					ctx, cancelCtx := context.WithTimeout(
+						context.Background(),
+						100*time.Millisecond,
+					)
+					defer cancelCtx()
+
+					readyMembersIndexes, err := announcer.announce(
+						ctx,
+						memberIndex,
+						test.message,
+						1,
+					)
+					if err != nil {
+						errorsChan <- &memberError{memberIndex, err}
+						return
+					}
+
+					resultsChan <- &memberResult{memberIndex, readyMembersIndexes}
+				}(broadcastingMemberIndex)
+			}
+
+			wg.Wait()
+
+			close(resultsChan)
+			results := make(map[group.MemberIndex][]group.MemberIndex)
+			for r := range resultsChan {
+				results[r.memberIndex] = r.readyMembersIndexes
+			}
+
+			close(errorsChan)
+			errors := make(map[group.MemberIndex]error)
+			for e := range errorsChan {
+				errors[e.memberIndex] = e.err
+			}
+
+			if !reflect.DeepEqual(test.expectedErrors, errors) {
+				t.Errorf(
+					"unexpected errors\n"+
+						"expected: [%v]\n"+
+						"actual:   [%v]",
+					test.expectedErrors,
+					errors,
+				)
+			}
+
+			if !reflect.DeepEqual(test.expectedResults, results) {
+				t.Errorf(
+					"unexpected results\n"+
+						"expected: [%v]\n"+
+						"actual:   [%v]",
+					test.expectedResults,
+					results,
+				)
+			}
 		})
 	}
 }
+
+type mockSigningAnnouncer struct {
+	outgoingAnnouncements map[uint64]struct {
```

Comment:
Maybe `performedAnnouncements` / `announcementsExecuted` / `announcementParams`?

Reply:
What's wrong with `outgoingAnnouncements`? Those are announcements that left the announcer which is similar to the real implementation of the announcer.

Reply:
I see now what you meant by `outgoingAnnouncements` and `incomingAnnouncementsFn`. I thought the first one is just capturing the parameters for the `announce` function calls so that we can later assert if this function was called with the right parameters. And for the second, I assumed this is just what we want to return, not what is incoming from the outside to the given announcer instance.

Can we add some code comments explaining these two fields?